freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/conversion/UserMetadataConverter.java

```diff
@@ -0,0 +1,100 @@
+        requireNonNull(user);
+
+        if (StringUtils.isBlank(user.getTitle())) {
+            // No need to warn, because this can happen under normal circumstances, if the FreeIPA user record was created before the credential
+            // update optimization entitlement was granted to the account.
+            return Optional.empty();
+        }
+
+        JsonUserMetadata jsonUserMetadata;
+        try {
+            jsonUserMetadata = OBJECT_READER.readValue(user.getTitle());
+        } catch (JsonProcessingException e) {
+            LOGGER.warn("Unable to parse metadata string {} for user with workload username {}", user.getTitle(), user.getUid(), e);
+            return Optional.empty();
+        }
+
+        Crn crn;
+        try {
+            crn = Crn.safeFromString(jsonUserMetadata.getCrn());
+        } catch (CrnParseException e) {
+            LOGGER.warn("Unable to parse CRN {} in metadata for user with workload username {}", jsonUserMetadata.getCrn(), user.getUid(), e);
+            return Optional.empty();
+        }
+        if (crn.getService() != Crn.Service.IAM ||
+                (crn.getResourceType() != Crn.ResourceType.USER && crn.getResourceType() != Crn.ResourceType.MACHINE_USER)) {
+            LOGGER.warn("CRN {} in metadata for user with workload username {} is not a user or machine user CRN", crn, user.getUid());
+            return Optional.empty();
+        }
+
+        UserManagementProto.WorkloadUserSyncActorMetadata protoMetadata;
+        try {
+            byte[] base64Decoded = BaseEncoding.base64().decode(jsonUserMetadata.getMeta());
+            protoMetadata = UserManagementProto.WorkloadUserSyncActorMetadata.parseFrom(base64Decoded);
+        } catch (Exception e) {
+            LOGGER.warn("Unable to decode meta attribute {} for user work workload username {}", jsonUserMetadata.getMeta(), user.getUid(), e);
+            return Optional.empty();
+        }
+
+        return Optional.of(new UserMetadata(jsonUserMetadata.getCrn(), protoMetadata.getWorkloadCredentialsVersion()));
+    }
+
+    public String toUserMetadataJson(UserMetadata userMetadata) {
+        requireNonNull(userMetadata);
+        return toUserMetadataJson(userMetadata.getCrn(), userMetadata.getWorkloadCredentialsVersion());
+    }
+
+    public String toUserMetadataJson(String crn, long workloadCredentialsVersion) {
+        checkArgument(!Strings.isNullOrEmpty(crn));
+
+        UserManagementProto.WorkloadUserSyncActorMetadata protoMetadata = UserManagementProto.WorkloadUserSyncActorMetadata.newBuilder()
+                .setWorkloadCredentialsVersion(workloadCredentialsVersion)
+                .build();
+        String protoMetadataBase64Encoded = BaseEncoding.base64().encode(protoMetadata.toByteArray());
+
+        JsonUserMetadata jsonUserMetadata = new JsonUserMetadata(crn, protoMetadataBase64Encoded);
+        try {
+            return OBJECT_WRITER.writeValueAsString(jsonUserMetadata);
```

Comment:
This is the point where I would say a comment would be nice to have some details about the decision why we have this complicated logic to store a long. I know there is the jira with the doc, but 2 lines should be sufficient. thanks

Reply:
Done.

Also, with regard to this comment on ll. 35-38 of UserMetadataConverter:

> please remove @SuppressWarnings and refactor this method. I agree with @handavid , it's hard to follow what happens here

This is done too. (Github wouldn't let me reply directly to the comment.)